### PR TITLE
feat(Core/GameObject): Expand IsSummonedBy() to GameObjects

### PR DESCRIPTION
**Pull request by Nyeriah**
cherry-pick: https://github.com/TrinityCore/TrinityCore/commit/5f545f540216d3b94c26e4aeda50c8bb8d5c3d74#

## Changes Proposed:
-  Allows gameobjects to be summon owners/expands the IsSummonedBy() event hook to accept gameobjects
-  Adds some other hooks for gameobject ai (SummonedCreatureDies, etc)

## Issues Addressed:

- Closes 

## Tests Performed:

- 
- 

## How to Test the Changes:

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Creature/TemporarySummon.cpp [hunk 1]
```diff
@@ -291,8 +305,18 @@ void TempSummon::UnSummon(uint32 msTime)
     }
 
     Unit* owner = GetSummonerUnit();
```
Comment (Kitzunu) [inline src/server/game/Entities/Creature/TemporarySummon.cpp:307]: unused

Comment (Kitzunu) [reply src/server/game/Entities/Creature/TemporarySummon.cpp:307]:
```suggestion
```

File: src/server/game/Entities/Creature/TemporarySummon.cpp [hunk 2]
```diff
@@ -290,9 +304,18 @@ void TempSummon::UnSummon(uint32 msTime)
         return;
     }
 
-    Unit* owner = GetSummonerUnit();
-    if (owner && owner->GetTypeId() == TYPEID_UNIT && owner->ToCreature()->IsAIEnabled)
-        owner->ToCreature()->AI()->SummonedCreatureDespawn(this);
+
```
Comment (Nyeriah) [inline src/server/game/Entities/Creature/TemporarySummon.cpp:307]:
```suggestion
```

**Comment by Nyeriah** [comment]
fix eluna...
https://github.com/azerothcore/mod-eluna/pull/89/files

**Review by Kitzunu — APPROVED** [review]

**Pull request merged**